### PR TITLE
BUG/DOC/TST: combine_pvalues: fix Tippett and Pearson

### LINKED ISSUE #15373: BUG: Tippett’s and Pearson’s method for combine_pvalues are not monotonous
### Describe your issue.

Something is wrong with the methods `pearson` and `tippett` for `scipy.stats.combine_pvalues`. They should be monotonically increasing with all components of the input, yet they are monotonically decreasing. At first glance, the output seems to be 1 − _p_ when it should be _p._ It’s also straightforward to compute that Tippett’s method for an input of (½,½) should yield _p_ = ¾ but it does not.

@rlucas7 IIUC, you should know best about this.

### Reproducing Code Example

```python
import numpy as np
from scipy.stats import combine_pvalues

n = 5
ps = np.linspace(0.1,0.9,5)

for method in ("pearson","tippett"):
	combined_ps = [
			combine_pvalues( np.full(n,p), method )[1]
			for p in ps
		]
	assert np.all( np.diff(combined_ps) >= 0 )

assert combine_pvalues([0.5,0.5],method="tippett") == 0.75
```

### Error message

All assertions fail.

### SciPy/NumPy/Python version information

1.6.3 1.19.5 sys.version_info(major=3, minor=9, micro=7, releaselevel='final', serial=0)

### PR DESCRIPTION
**Pull request by Wrzlprmft**
#### Reference issue
Closes #15373.

#### What does this implement/fix?
* Fixes #15373 and includes test for monotony as described therein.
* Includes warnings in documentation that this can only be applied to continuous tests.
* Better explain the different methods in the documentation, removing technical details that I doubt to be useful to any reader.
* Minor formatting fixes to documentation.

#### Additional information
I am currently in the process of writing a Monte Carlo-based [module](https://github.com/BPSB/combine-p-values-discrete) for combining _p_ values (mainly of discrete tests): I used this to further test all methods.

### CODE REVIEW

**Comment by AtsushiSakai** [comment]
There are some lint errors:
```
scipy/stats/_stats_py.py:8118:80: E501 line too long (87 > 79 characters)
scipy/stats/_stats_py.py:8196:80: E501 line too long (98 > 79 characters)
```
[Link](https://dev.azure.com/scipy-org/SciPy/_build/results?buildId=16086&view=logs&jobId=fdd89725-f657-5341-7d77-3c5c44c4f5b6&j=fdd89725-f657-5341-7d77-3c5c44c4f5b6&t=325fc03e-e0b9-5339-adaf-099a49a14600)

**Comment by Wrzlprmft** [comment]
> There are some lint errors:

I know, but I don’t know how to (reasonably) fix either of them: The first is a long URL that cannot be wrapped to 80 chars. Even if this could be wrapped somehow, docstring readability would inevitably suffer. The second is an overly long parameter description that does not get appropriately rendered when wrapped, or at least I don’t know how. The first one is also not by me, but the lintdiff doesn’t recognise that it is.

Ceterum censeo: Manual wrapping of texts is an annoying archaism: It breaks everything once you view something at a smaller width and computers can perfectly do it automatically since decades for every width.

File: scipy/stats/_stats_py.py [hunk 1]
```diff
@@ -8107,26 +8107,27 @@ def brunnermunzel(x, y, alternative="two-sided", distribution="t",
...
 
     Parameters
     ----------
     pvalues : array_like, 1-D
-        Array of p-values assumed to come from independent tests.
-    method : {'fisher', 'pearson', 'tippett', 'stouffer',
-              'mudholkar_george'}, optional
+        Array of p-values assumed to come from independent tests based on
+        continuous distributions.
+    method : {'fisher', 'pearson', 'tippett', 'stouffer', 'mudholkar_george'}, optional
```
Comment (tupui) [inline scipy/stats/_stats_py.py:8118]: Please revert the change on `method` as the formatting is correct. Another alternative would be to remove `optional` since the first parameter is the default in numpydoc terms.
```suggestion
    method : {'fisher', 'pearson', 'tippett', 'stouffer', 'mudholkar_george'}
```

File: scipy/stats/_stats_py.py [hunk 2]
```diff
@@ -8139,44 +8140,67 @@ def combine_pvalues(pvalues, method='fisher', weights=None):
...
+    .. [1] Kincaid, W. M., "The Combination of Tests Based on Discrete
+           Distributions." Journal of the American Statistical Association 57,
+           no. 297 (1962), 10-19.
+    .. [2] Heard, N. and Rubin-Delanchey, P. "Choosing between methods of
            combining p-values."  Biometrika 105.1 (2018): 239-246.
-    .. [5] Whitlock, M. C. "Combining probability from independent tests: the
+    .. [3] https://en.wikipedia.org/wiki/Fisher%27s_method
+    .. [4] George, E. O., and G. S. Mudholkar. "On the convolution of logistic
+           random variables." Metrika 30.1 (1983): 1-13.
+    .. [5] https://en.wikipedia.org/wiki/Fisher%27s_method#Relation_to_Stouffer.27s_Z-score_method
```
Comment (tupui) [inline scipy/stats/_stats_py.py:8196]: Fine to get a lint error for links.

File: scipy/stats/_stats_py.py [hunk 3]
```diff
@@ -8215,8 +8239,8 @@ def combine_pvalues(pvalues, method='fisher', weights=None):
 
     else:
         raise ValueError(
-            "Invalid method '%s'. Options are 'fisher', 'pearson', \
-            'mudholkar_george', 'tippett', 'or 'stouffer'", method)
+            "Invalid method '%s'. Valid methods are 'fisher', 'pearson', \
+            'mudholkar_george', 'tippett', and 'stouffer'", method)
```
Comment (tupui) [inline scipy/stats/_stats_py.py:8243]: Since you touch this, let's make it pretty
```suggestion
            f"Invalid method {method!r}. Valid methods are 'fisher', 'pearson', "
            "'mudholkar_george', 'tippett', and 'stouffer'"
        )
```

File: scipy/stats/tests/test_stats.py [hunk 4]
```diff
@@ -6615,6 +6615,23 @@ def test_mudholkar_george_equal_fisher_minus_pearson(self):
         # 0.5 here is because logistic = log(u) - log(1-u), i.e. no 2 factors
         assert_approx_equal(0.5 * (Z_f-Z_p), Z, significant=4)
 
+    @pytest.mark.parametrize("variant", ["single", "all"])
+    @pytest.mark.parametrize("method",
+                             ["fisher", "pearson", "tippett",
+                                 "stouffer", "mudholkar_george"])
+    def test_monotony(self, variant, method):
```
Comment (tupui) [inline scipy/stats/tests/test_stats.py:6622]:
```suggestion
    @pytest.mark.parametrize("method",
                             ["fisher", "pearson", "tippett",
                              "stouffer", "mudholkar_george"])
    def test_monotony(self, variant, method):
```

File: scipy/stats/tests/test_stats.py [hunk 5]
```diff
@@ -6615,6 +6615,23 @@ def test_mudholkar_george_equal_fisher_minus_pearson(self):
...
+        # Test that result increases monotonously with respect to input.
+        changing_values = np.linspace(0.1, 0.9, 10)
+        pvalues = np.random.random(7)
```
Comment (tupui) [inline scipy/stats/tests/test_stats.py:6625]: We want reproducible results
```suggestion
        rng = np.random.default_rng()
        pvalues = rng.random(7)
```

**Review by tupui — CHANGES_REQUESTED** [review]
Thanks @Wrzlprmft. Regarding line length, aside from urls we don't compromise. I disagree with the line wrapping argument. There are many reasons you would not want to rely on this (basic terminal, weird wrapping formatting, consistency throughout the code base, etc.)

@steppi since you had a close look at the issue, could you review this?

(Note: we will probably squash unless there is a history re-write)

File: scipy/stats/_stats_py.py (hunk 1 above)
Comment (Wrzlprmft) [reply scipy/stats/_stats_py.py:8118]: > Please revert the change on `method` as the formatting is correct.

Well, it doesn’t [get properly rendered](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.combine_pvalues.html). Removing the `optional` is fine by me though.

Comment (tupui) [reply scipy/stats/_stats_py.py]: Indeed, I will have a look at similar cases. It should work or there is a small tweak I don't see here.

File: scipy/stats/tests/test_stats.py (hunk 5 above)
Comment (Wrzlprmft) [reply scipy/stats/tests/test_stats.py:6625]: Would we not have to specify a seed to make this really reproducible? Or does the test environment somehow ensure for a reproducible seed here?

File: scipy/stats/_stats_py.py (hunk 1 above)
Comment (tupui) [reply scipy/stats/_stats_py.py]: I am sure it was working before, I see other places where this is broken as well. Thanks for noticing this.

File: scipy/stats/tests/test_stats.py (hunk 5 above)
Comment (tupui) [reply scipy/stats/tests/test_stats.py:6625]: We don't rely on global states like `np.random.seed` as this is not recommended anymore. All new code should use explicit seeding of the `np.random.Generator`. Also in this case, a global state would not be reliable as the generated numbers would be different if you launch tests in a different order (which happens all the time). Test must be independent.

Comment (Wrzlprmft) [reply scipy/stats/tests/test_stats.py:6625]: Sure, but wouldn’t we need to do something like `RNG = np.random.default_rng(42)` then? Because in your suggestion, `default_rng` is globally seeded or even machine-state-seeded, unless I am missing something.

Comment (tupui) [reply scipy/stats/tests/test_stats.py:6625]: Oh yes totally, it's a typo on my side. Fixing the suggestion.

File: scipy/stats/_stats_py.py (hunk 2 above)
Comment (Wrzlprmft) [reply scipy/stats/_stats_py.py:8196]: Also note that there were some broken references, e.g., the last paragraph referenced Ref. 5, when it should be referencing Ref. 7. Without reordering, things would be all over the place.

File: scipy/stats/_stats_py.py [hunk 6]
```diff
@@ -8215,8 +8239,9 @@ def combine_pvalues(pvalues, method='fisher', weights=None):
 
     else:
         raise ValueError(
-            "Invalid method '%s'. Options are 'fisher', 'pearson', \
-            'mudholkar_george', 'tippett', 'or 'stouffer'", method)
+            f"Invalid method {method!r}. Valid methods are 'fisher', 'pearson', "
+            "'mudholkar_george', 'tippett', and 'stouffer'"
+        )
```
Comment (tupui) [inline scipy/stats/_stats_py.py:8244]:
```suggestion
            f"Invalid method {method!r}. Valid methods are 'fisher', "
            "'pearson', 'mudholkar_george', 'tippett', and 'stouffer'"
        )
```

**Review by Wrzlprmft — COMMENTED** [review]
(Interface is weird here, not sure if I am reviewing myself now or can do anything at all:) I approve of these changes.

**Comment by steppi** [comment]
> @steppi since you had a close look at the issue, could you review this?

Sure thing. Curiously, the project I was working on that lead to my first PR to SciPy, gh-13979, involved implementing and comparing different methods for combining p-values when working with only the log p-values.

File: scipy/stats/tests/test_stats.py [hunk 7]
```diff
@@ -6615,6 +6615,24 @@ def test_mudholkar_george_equal_fisher_minus_pearson(self):
...
+        pvalues = rng.random(7)
+        combined_pvalues = []
+        for changing_value in changing_values:
+            if variant == "single":
+                pvalues[0] = changing_value
+            else:
+                pvalues = np.full(7, changing_value)
+            combined_p = stats.combine_pvalues(pvalues, method=method)[1]
+            combined_pvalues.append(combined_p)
+        assert np.all(np.diff(combined_pvalues) >= 0)
```
Comment (steppi) [inline scipy/stats/tests/test_stats.py:6635]: I found the logic here a little confusing. For the "all" option, it's generating a set of p_values but then
immediately overwriting them with a constant array. It ends up just testing with an increasing sequence of constant arrays. What do you think of the following modification? It generates an increasing sequence of (non-constant) arrays all at once and then replaces the loop with a list comprehension. It leaves out the "single" option, but I'm not sure this option is actually necessary.

```suggestion
        m, n = 10, 7
        # Generate an m x n array of p_values. Each row corresponds to
        # a set of p_values to be combined with p_values increasing
        # monotonically down each column.
        p_values_array = np.sort(rng.uniform(0, 1, size=(m, n)), axis=0)
        # Generate combined p_values for each row in array.
        combined_pvalues = [
            stats.combine_pvalues(p_values_array[i, :], method=method)[1]
            for i in range(m)
        ]
        # Check that combined p_values increase monotonically as expected.
        assert np.all(np.diff(combined_pvalues) >= 0)
```

**Review by steppi — CHANGES_REQUESTED** [review]
I finally had some time to review this. The actual bug has indeed been fixed.  I like the idea of cautioning users that these methods produce systematically biased results for inherently discrete tests. I've actually run into that problem before while trying to combine p_values based on permutation tests. I ended up having to search to find out why the results were strange. I've made some suggestions for the docstring and the tests, but nothing too major.

File: scipy/stats/tests/test_stats.py (hunk 7 above)
Comment (Wrzlprmft) [reply scipy/stats/tests/test_stats.py]: I updated this as I described above. For each of the three variants the test data is created separately (and hopefully clearly).

File: scipy/stats/tests/test_stats.py [hunk 8]
```diff
@@ -6615,6 +6615,24 @@ def test_mudholkar_george_equal_fisher_minus_pearson(self):
...
         assert_approx_equal(0.5 * (Z_f-Z_p), Z, significant=4)
 
+    @pytest.mark.parametrize("variant", ["single", "all"])
+    @pytest.mark.parametrize("method",
+                             ["fisher", "pearson", "tippett",
+                              "stouffer", "mudholkar_george"])
+    def test_monotony(self, variant, method):
+        # Test that result increases monotonously with respect to input.
+        changing_values = np.linspace(0.1, 0.9, 10)
+        rng = np.random.default_rng(278448169958891062669391462690811630763)
```
[the remark this answers is not in the document]
Comment (steppi) [reply scipy/stats/tests/test_stats.py:6626]: Oh, sorry I missed that. I had received a conflicting suggestion before and was passing it on. I didn't realize that the recommended way had changed.

File: scipy/stats/_stats_py.py [hunk 9]
```diff
@@ -8139,44 +8140,67 @@ def combine_pvalues(pvalues, method='fisher', weights=None):
...
+      averaging their statistics [4]_. Their method emphasises extreme
+      p-values, both close to 1 and 0.
+    * Stouffer's method [5]_ uses Z-scores and the statistic:
+      :math:`\\sum_i \\Phi^{-1} (p_i)`, where :math:`\\Phi` is the CDF of the
+      standard normal distribution. The advantage of this method is that it is
+      straightforward to introduce weights, which can make Stouffer's method
+      more powerful than Fisher's method when the p-values are from studies
+      of different size [6]_ [7]_.
+    * Tippett's method uses the smallest p-value as a statistic.
+      (Mind that this minimum is not the combined p-value.)
```
[the remark this answers is not in the document]
Comment (steppi) [reply scipy/stats/_stats_py.py:8188]: I see, you make good points. If we're going with the updated version I have quibbles but won't have time to make them until later today.

File: scipy/stats/_stats_py.py [hunk 10]
```diff
@@ -8139,44 +8150,67 @@ def combine_pvalues(pvalues, method='fisher', weights=None):
...
-    `mudholkar_george` method is the difference of the Fisher's and Pearson's
-    test statistics, each of which include the -2 factor [4]_. However, the
-    `mudholkar_george` method does not include these -2 factors. The test
-    statistic of `mudholkar_george` is the sum of logisitic random variables and
-    equation 3.6 in [3]_ is used to approximate the p-value based on Student's
-    t-distribution.
+    All methods assume tests based on continuous distributions, i.e., the
+    distribution of p-values under the null hypothesis must be the continuous
+    uniform distribution on the interval [0,1]. If this function is applied to
+    tests with a discrete statistics such as any rank test or contingency-table
```
Comment (steppi) [inline scipy/stats/_stats_py.py:8156]: This brings up a subtlety. Among contingency table tests, something like Fisher's exact test is very obviously discrete, but in principle, a 𝝌² test for contingency tables is based upon a continuous distribution, but for smaller sample sizes the distribution of p-values is discrete because the data is discrete. Since your not trying to be exhaustive with your examples maybe just use a more obviously discrete family of tests like permutation tests of which Fisher's exact test is also an example.

```suggestion
    tests with a discrete statistics such as any rank test or permutation
```

File: scipy/stats/_stats_py.py [hunk 11]
```diff
@@ -8139,44 +8150,67 @@ def combine_pvalues(pvalues, method='fisher', weights=None):
...
+    when considering significance [2]_. For example, methods emphasising large
+    p-values are more sensitive to strong false and true negatives; conversely
+    methods focussing on small p-values are sensitive to positives.
+
+    * Fisher's method (also known as Fisher's combined probability test) [3]_
+      uses the product of individual p-values: :math:`\\prod_i p_i`.
+      (Mind that this product is not the combined p-value.)
+      This method emphasises small p-values.
+      Note that the test statistic used internally and returned by this method
+      is :math:`-2\\sum_i \\log(p_i)` for numerical reasons.
```
Comment (steppi) [inline scipy/stats/_stats_py.py:8173]: I think it's actually of interest that this quantity follows a 𝝌² distribution. It's the historical reason why Fisher's exact test was adopted, despite having some defects such as being sensitive to very small p-values (as you point out). I think you should still mention it somewhere, otherwise this formula seems unmotivated.

**Review by steppi — CHANGES_REQUESTED** [review]
I've made some quibbles about the Docstring. Btw, I like how you've updated the test.

File: scipy/stats/tests/test_stats.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (tupui) [reply scipy/stats/tests/test_stats.py:6626]: For testing purposes, there is nothing wrong per se with 1234. It's yet again a way to prevent users to copy paste seeds in their codes. Note that from the RNG perspective, small seeds are bad for initialization. This is why NumPy recommends using `np.random.SeedSequence().entropy` when there is a need for a seed.

[the remark this answers is not in the document]
Comment (steppi) [reply scipy/stats/tests/test_stats.py:6626]: I see, thanks for the explanation, it makes sense. I've never actually looked into the details of how NumPy's random number generators work. I think some will run a fixed number of iterations from the initial seed before using the output to get around such issues, although maybe that's still not good enough. For actual research practice I have one concern though: how are we to know that researchers have actually used `np.random.SeedSequence().entropy` properly rather than for example using it repeatedly until say their proposed machine learning algorithm achieves state of the art results on a set of benchmarks? Obviously that's not a concern for this PR though.

[the remark this answers is not in the document]
Comment (tupui) [reply scipy/stats/tests/test_stats.py:6626]: > how are we to know that researchers have actually used np.random.SeedSequence().entropy properly rather than for example using it repeatedly until say their proposed machine learning algorithm achieves state of the art results on a set of benchmarks? 

That's exactly what I don't like about being able to seed 😉 I've seen a lot of academic or production code with generators fixed to a given seed. Then you cannot say anymore that you are using a random number generator, it's just a fixed sequence which seems to be random.

Personally, I prefer to differentiate random generators and sequences. For cases where you need to have some reproducibility of the stream of number, I would use low discrepancy sequence (because the stream is of better quality than a MC with a fixed seed and also not platform/implementation dependent, etc.) and when you need random numbers then forbid seeding of MC or use scrambled sequences. This way it's clearer and you're less prone to error/bias.

But to your question, in the end, people will always find a way to manipulate the results to their advantage... 😅

[the remark this answers is not in the document]
Comment (steppi) [reply scipy/stats/tests/test_stats.py:6626]: > hat's exactly what I don't like about being able to seed wink I've seen a lot of academic or production code with generators fixed to a given seed. Then you cannot say anymore that you are using a random number generator, it's just a fixed sequence which seems to be random.

Interesting. There's one case where I've used an equidistributed sequence in a quasi-monte-carlo algorithm, but I'd never considered them as a replacement for psuedorandom numbers before.  For randomized algorithms, I think what one actually desires is a fixed sequence which seems to be random. Von Neumann had this to say on the matter:

*The real objection to this procedure is the practical need for checking computations. If we suspect that a calculation is wrong, almost any reasonable check involves repeating something done before. At that point the generation of new random numbers would be intolerable.* [1]

There's also cases where using a fixed sequence reduces unwanted variability. Such as when selecting cross-validation splits for a data-set. Reproducibility is really important. I guess I need to look into the literature more on how different randomized algorithms are affected when using a low discrepancy sequence. I'd be really cautious about the idea of ever using non-reproducible sequences of random numbers though.



[1] Von Neumann, John (1951). ["Various techniques used in connection with random digits"](https://dornsifecms.usc.edu/assets/sites/520/docs/VonNeumann-ams12p36-38.pdf) (PDF). National Bureau of Standards Applied Mathematics Series. 12: 36–38.

[the remark this answers is not in the document]
Comment (tupui) [reply scipy/stats/tests/test_stats.py:6626]: Ah yes, I remember this article. Thanks for reminding me of it, it's a great read!

> There's also cases where using a fixed sequence reduces unwanted variability. Such as when selecting cross-validation splits for a data-set. Reproducibility is really important.

I think that this is a good use case for low discrepancy sequences. They provide reproducible sequences of numbers but properly sample the parameter space (this is even worse for MC if you would sample in n-D), which is what you want when splitting data-set. With MC you have a lot of variance in the quality of the sample and could pick a very good seed or a very bad one from one seed to another. Using sequences also helps a lot if you really care about reproducibility since the stream of numbers is defined and known.

Here I am comparing QMC sequences to things like `np.random.default_rng(0)` or `np.random.default_rng(1234)` which are widely used.

[the remark this answers is not in the document]
Comment (steppi) [reply scipy/stats/tests/test_stats.py:6626]: > I think that this is a good use case for low discrepancy sequences.

Yeah, I already agree there actually. Von Neumann goes on to say in that paper that being able to check calculations in case of mistakes is the **only** reason to prefer reproducible sequences, so I just wanted to cover my bases with a class of reasons he didn’t foresee.

[the remark this answers is not in the document]
Comment (steppi) [reply scipy/stats/tests/test_stats.py:6626]: I’m really curious how well something like a random forest would perform with psuedorandom numbers vs a low-discrepancy sequence. My intuition says that pseudorandom would work better, but I want to check over the weekend.

[the remark this answers is not in the document]
Comment (tupui) [reply scipy/stats/tests/test_stats.py:6626]: I am interested by things like this, if you could keep me updated 😃 I am part of the organization of a QMC conference (MCQMC2022) and this is something that could be of interest.

File: scipy/stats/_stats_py.py (hunk 10 above)
Comment (Wrzlprmft) [reply scipy/stats/_stats_py.py:8156]: I would leave this at contingency tables since they are the more important example in application, discrete in the respect that matters here, and easily discernible as discrete by the target audience for this information. Whether some continuous distribution is used somewhere under the hood (e.g., as an estimator) does not affect this. Also, all tests approximate continuity for a sufficiently large sample size.

File: scipy/stats/_stats_py.py [hunk 12]
```diff
@@ -8139,44 +8150,67 @@ def combine_pvalues(pvalues, method='fisher', weights=None):
...
+
+    The differences between the methods can be best illustrated by their
+    statistics and what aspects of a combination of p-values they emphasise
+    when considering significance [2]_. For example, methods emphasising large
+    p-values are more sensitive to strong false and true negatives; conversely
+    methods focussing on small p-values are sensitive to positives.
+
+    * Fisher's method (also known as Fisher's combined probability test) [3]_
+      uses the product of individual p-values: :math:`\\prod_i p_i`.
+      (Mind that this product is not the combined p-value.)
```
[the remark this answers is not in the document]
Comment (Wrzlprmft) [reply scipy/stats/_stats_py.py:8170]: That’s a valid point. I streamlined this (and analogous parts).

File: scipy/stats/_stats_py.py [hunk 13]
```diff
@@ -8139,44 +8150,67 @@ def combine_pvalues(pvalues, method='fisher', weights=None):
...
-    quantity has a chi-square distribution that determines the p-value. The
-    `mudholkar_george` method is the difference of the Fisher's and Pearson's
-    test statistics, each of which include the -2 factor [4]_. However, the
-    `mudholkar_george` method does not include these -2 factors. The test
-    statistic of `mudholkar_george` is the sum of logisitic random variables and
-    equation 3.6 in [3]_ is used to approximate the p-value based on Student's
-    t-distribution.
+    All methods assume tests based on continuous distributions, i.e., the
+    distribution of p-values under the null hypothesis must be the continuous
+    uniform distribution on the interval [0,1]. If this function is applied to
```
[the remark this answers is not in the document]
Comment (Wrzlprmft) [reply scipy/stats/_stats_py.py:8155]: I think the test statistics is secondary here. The distorting effect arises from only discrete *p* values being possible, whether it’s due to the discrete nature of the data (which is the most usual case) or due to some approximation used that can only yield discrete results (which should not be relevant today anymore). At the end of the day, we are only describing the null hypothesis of these methods. If somebody (as in your meta-analysis scenario) knowingly feeds these methods data based on different assumptions, they should also be able to adopt these aspects accordingly.

File: scipy/stats/_stats_py.py (hunk 11 above)
Comment (Wrzlprmft) [reply scipy/stats/_stats_py.py:8173]: I added this aspect back, however not that prominently and only after introducing the test

Also, there is a straightforward motivation for Fisher’s method beyond historical practical considerations: If you want to know how unlikely two events are to coincide, you multiply their probabilities.

**Comment by Wrzlprmft** [comment]
@steppi: I streamlined the docstring once more, including your comments. In addition, I removed the redundancy of the methods list in the arguments listing all the test statistics. It now only spells out the names of all the methods. I think the detailed information about the different methods only needs to be in one place and that are is the *Notes.*

**Comment by Wrzlprmft** [comment]
@steppi @tupui I also noted that the statistics for Pearson’s method has an odd sign in context. Although there is no clear consensus on what to use as a statistics for Pearson’s method, most sources that compare methods give statistics for both Fisher’s and Pearson’s method such that a higher statistics corresponds to a lower *p* value and Mudholkar’s and George’s method averages Fisher’s and Pearson’s statistics (or is their sum). See for example Heard and Rubin-Delanchy (2017, referenced in the docstring).

I therefore have also changed the sign of Pearson’s statistics. The clear drawback of this is that it’s backwards-incompatible, but then Pearson’s method was erroneous until this pull request anyway.

File: scipy/stats/_stats_py.py (hunk 10 above)
Comment (steppi) [reply scipy/stats/_stats_py.py]: Yeah. This suggestion isn't strict. It's OK to disregard. As someone trained as a mathematician, I'm kind of a stickler about precise language. It's a matter of how one defines *based upon a continuous distribution.* Is it that the test statistic follows a continuous distribution in theory, or that the statistics available in practice take values in something approximating a continuum? Clearly you're using the second definition, and hopefully people will find it clear from the context. Also, I guess deep down everything in computational statistics is discrete.

File: scipy/stats/_stats_py.py (hunk 11 above)
Comment (steppi) [reply scipy/stats/_stats_py.py]: > Also, there is a straightforward motivation for Fisher’s method beyond historical practical considerations: If you want to know how unlikely two events are to coincide, you multiply their probabilities.

I'm not so sure. This is what Heard and Rubin-Delanchy had to say on the matter

*However, little attention has been afforded to highlighting precisely the settings in which these
combiners are optimal, and thereby determining their suitability for an application at hand.
Instead, the widespread adoption of the above statistics for combining p-values can be
largely attributed to their simplicity and mathematical convenience: Under H0, −2SP and
−2SF are both distributed as χ2 2n; SG and SE have slightly more awkward closed-form
distributions which can be well approximated by Gaussian distributions for large n; SS
is N(0, n); and ST ∼ Beta(1, n). A meta-analysis combined p-value is therefore trivial
to obtain for each of these statistics, but they can differ substantially.*

Also I think explaining Fisher's method by saying if you want to know how unlikely two events are to coincide, you
multiply their probabilities is one of those intuitive but incomplete explanations common in the sciences. With
meta-analysis methods like these, what one typically wants is to be able to aggregate p-values from multiple studies, which may themselves be weakly powered to get a result that is more strongly powered. It does not appear obvious to me that the probability that the null hypothesis was incorrectly rejected for every test is the best statistic for performing meta-analysis of multiple studies. I really think the reason Fisher's method became among competing methods is due to its ease of computation to the the nice way the distribution works out.

**Review by steppi — APPROVED** [review]
@Wrzlprmft I'm happy with the docstring now. Thanks for putting in the effort. Modulo debate about the sign of the test statistic, I think this looks good and so am approving.

File: scipy/stats/_stats_py.py (hunk 11 above)
Comment (Wrzlprmft) [reply scipy/stats/_stats_py.py]: No disagreement here, hence *motivation* and not *watertight rationale.* There are clear downsides to Fisher’s method, such as completely ignoring if some studies point significantly in the other direction.

**Comment by steppi** [comment]
> I therefore have also changed the sign of Pearson’s statistics. The clear drawback of this is that it’s backwards-incompatible, but then Pearson’s method was erroneous until this pull request anyway.

Personally I prefer having the sign the way you've done it, but defer to @tupui as to whether it's worth breaking backwards compatibility.

**Comment by tupui** [comment]
I would suspect that the statistic is less used than the p-value, so there should be a limited impact. In case of a clear bug, we should fix it.

**Comment by Wrzlprmft** [comment]
> In case of a clear bug, we should fix it.

@tupui That’s the conundrum: It’s not a clear bug. There are several possible choices for the statistics for this method, which are all easily transformable into each other. The one chosen before is just unusual, in particular in the context of the other methods. However, the *p* value produced by the method in question was blatantly wrong and fixing that is one of the most important parts of this pull request. Thus, the only people who might stumble upon this sign change would be those who use the method only for the statistic, which I suspect are none because I don’t see a reason to further process it and it’s straightforward to compute (in whatever variant you want).

**Comment by steppi** [comment]
> > In case of a clear bug, we should fix it.
> 
> @tupui That’s the conundrum: It’s not a clear bug. There are several possible choices for the statistics for this method, which are all easily transformable into each other. The one chosen before is just unusual, in particular in the context of the other methods. However, the _p_ value produced by the method in question was blatantly wrong and fixing that is one of the most important parts of this pull request. Thus, the only people who might stumble upon this sign change would be those who use the method only for the statistic, which I suspect are none because I don’t see a reason to further process it and it’s straightforward to compute (in whatever variant you want).

Although perhaps not a clear bug, I agree with you that it would be good to have the returned statistic be consistent with what is typically seen in the literature. I think any hypothetical person using this method just for the statistic wold have to be aware of the bug in the p-value calculation and will be happy that it has been fixed.

**Comment by tupui** [comment]
Let's do this then 👍

**Review by tupui — APPROVED** [review]
> Is there anything left for me to do? @tupui, you are still listed as requesting changes, but I think I addressed everything.

Indeed, I missed that you already did what we discussed. Thanks for the ping. CI failures are not relevant (the linter is about a long URL which was already there).

Thanks everyone 😃

**Comment by tupui** [comment]
@steppi could you update the release notes?

File: scipy/stats/tests/test_stats.py [hunk 14]
```diff
@@ -6598,23 +6598,49 @@ def test_weighted_stouffer2(self):
...
+    @pytest.mark.parametrize(
+        "method",
+        ["fisher", "pearson", "tippett", "stouffer", "mudholkar_george"],
+    )
+    def test_monotonicity(self, variant, method):
+        # Test that result increases monotonically with respect to input.
+        m, n = 10, 7
+        rng = np.random.default_rng(278448169958891062669391462690811630763)
+
+        # `pvaluess` is an m × n array of p values. Each row corresponds to
```
Comment (mdhaber) [inline scipy/stats/tests/test_stats.py:6627]: This `×` is now causing a lint failure in master. It needs to be added to the list of special characters in [tools/unicode-check.py](https://github.com/scipy/scipy/pull/12694/files#diff-80f6ab2c49611d1d905daa01f95b04bb36e1ffeae092666b5cee52d27dbb312f).

Comment (tupui) [reply scipy/stats/tests/test_stats.py:6627]: ooops thanks Matt! I can fix this tomorrow first thing if you did not do before me until then.

Comment (mdhaber) [reply scipy/stats/tests/test_stats.py:6627]: See gh-15606 : )

**Comment by steppi** [comment]
@tupui @Wrzlprmft 

I've updated the release notes [here](https://github.com/scipy/scipy/wiki/Release-note-entries-for-SciPy-1.9.0). I've pasted what I've written below. Let me know if you think there are any changes that should be made.

The methods "pearson" and "tippet" from scipy.stats.combine_pvalues have been fixed to return the correct p-values, resolving [#15373](https://github.com/scipy/scipy/issues/15373). The sign of the test statistic returned for the method "pearson" has been flipped so that higher values of the statistic now correspond to lower p-values, making the statistic more consistent with those of the other methods and with the majority of the literature. In addition, the documentation for scipy.stats.combine_pvalues has been expanded and improved.

**Comment by tupui** [comment]
Thanks @steppi it's a good description. I would just put it instead in the backward incompatible section (at least the sign change part).

**Comment by steppi** [comment]
> I would just put it instead in the backward incompatible section (at least the sign change part).

Good catch. Done!
